### ISSUE TITLE
NPE when test throws an Exception with null stacktrace

### ISSUE THREAD

**Issue by ScottG489** [jlink/jqwik#283]
## Testing Problem

The problem is that if an exception is thrown that doesn't have a stacktrace, [ErrorEquivalenceChecker.firstRelevantStackTraceElement](https://github.com/jlink/jqwik/blob/cc952102e3209a20c63e9b1cb6a7c3007a028b91/engine/src/main/java/net/jqwik/engine/properties/shrinking/ErrorEquivalenceChecker.java#L37) ends up throwing an NPE.

It's a pretty fair assumption that an Exception would always have a stacktrace, but it might not when the **exception is a mocked object**.

I have some tests which orchestrate a mock to throw an exception. If you're familiar with Mockito, something like this (sorry that it's so contrived):
```java
Thing mockedThing = mock(Thing.class);
Exception mockedException = mock(MyException.class);
doThrow(mockException).when(mockedThing).foobar();
someMethodThatCallsFoobarMethod(mockedThing);
```
`mockedThing.foobar()` is called from within `someMethodThatCallsFoobarMethod(mockedThing)` which throws the mocked exception. This bubbles up from the test and jqwik tries to handle it in the code linked to above and an NPE is thrown.

## Suggested Solution
The obvious solution is to work with the assumption that the Exception may have a null stacktrace. After taking a quick look, `ErrorEquivalenceChecker.firstRelevantStackTraceElement` already returns an `Optional`. So perhaps if the stacktrace is `null` an `Optional.empty()` could be returned.

What are your thoughts?

## Discussion

So of course my test shouldn't have been throwing an exception otherwise it would always fail, but I don't think it should be blowing up in jwqik and it made tracking down my error difficult. I'm really not sure what jqwik could do with an Exception with no stacktrace, though.

As an aside, I'd normally not mock an Exception, but I'm testing a bit of (hopefully) temporary hacky code and I needed to do some deep mocking.

**Comment by jlink** [comment]
I’d argue that one should never mock an exception, but on the other hand a test framework should never blow up either. In your example you could, it seems, as easily create a real MyException instance. Let me think about it for a while…

**Comment by jlink** [comment]
@ScottG489 My current feeling is that handling `Throwable.getStackTrace()` being null would give way to handling a lot of other potential null returns that shouldn't occur in the first place. Is there a good argument that this is a special case?

One more thought: Maybe it's the general handling/presentation of internal exceptions that's the problem and that could be improved. What was the error message you got when you ran into the error? What error message could have made the problem obvious to you?

**Comment by ScottG489** [comment]
Yeah I'm definitely on the same page. Over-handling use cases can get patchy and lend to a lot of spaghetti code.

My particular case was parsing an error message for some data to take an action using that data (i.e.  `mockException.getCause().getCause().getMessage()`). The lib I was using has some gaps and I needed to use this hacky solution. So naturally my test was a bit hacky as well. I was mocking an exception `cause` chain because the error message I needed to get at was 2 exceptions down. It's arguably better to have used real exceptions, and I definitely prefer to not use mocks whenever possible.

I think I'm still siding towards the lib being able to handle everything gracefully. Maybe it's too hard of a line, but I don't think it should matter what the user is doing, the framework should not blow up.

I definitely see a possibility of better handling/presentation. I don't think handling a specific case like I suggested above is usually the right action.

I think what may have contributed to it being a tricky error was that there was nothing pointing back to my code in the stacktrace jqwik printed. It's very disorienting not having a stacktrace leading back to your own code. But since the exception had no stacktrace, I don't think there's anything you'd be able to do about that.

When debugging jqwik I did notice that the `.toString()` of the mocked object was something like `Mock for MyThingException, hashCode: 882141659`. If this had been printed then it would have made it more obvious. Not sure if that would just be helpful in this use case and not more broadly, though.

Looking over the code again in and around `ErrorEquivalenceChecker`, I don't really see a good way of handling this more generally. Perhaps it's more acceptable to handle this situation more specifically since the class seems to be only for managing exceptions? You'd have the best feel for that though.

Finally, here's the exception I got:
<details>
<summary>Stacktrace</summary>

```
java.lang.NullPointerException
java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2022)
	at net.jqwik.engine.properties.shrinking.PropertyShrinker.withTimeout(PropertyShrinker.java:100)
	at net.jqwik.engine.properties.shrinking.PropertyShrinker.shrink(PropertyShrinker.java:90)
	at net.jqwik.engine.properties.shrinking.PropertyShrinker.shrink(PropertyShrinker.java:72)
	at net.jqwik.engine.properties.GenericProperty.shrink(GenericProperty.java:214)
	at net.jqwik.engine.properties.GenericProperty.shrinkAndCreateCheckResult(GenericProperty.java:183)
	at net.jqwik.engine.properties.GenericProperty.check(GenericProperty.java:80)
	at net.jqwik.engine.execution.CheckedProperty.check(CheckedProperty.java:67)
	at net.jqwik.engine.execution.PropertyMethodExecutor.executeProperty(PropertyMethodExecutor.java:156)
	at net.jqwik.engine.execution.PropertyMethodExecutor.executeMethod(PropertyMethodExecutor.java:135)
	at net.jqwik.engine.execution.PropertyMethodExecutor.lambda$executePropertyMethod$1(PropertyMethodExecutor.java:115)
	at net.jqwik.api.lifecycle.AroundPropertyHook.lambda$static$0(AroundPropertyHook.java:46)
	at net.jqwik.engine.execution.lifecycle.HookSupport.lambda$wrap$0(HookSupport.java:26)
	at net.jqwik.engine.hooks.lifecycle.PropertyLifecycleMethodsHook.aroundProperty(PropertyLifecycleMethodsHook.java:57)
	at net.jqwik.engine.execution.lifecycle.HookSupport.lambda$wrap$1(HookSupport.java:31)
	at net.jqwik.engine.execution.lifecycle.HookSupport.lambda$wrap$0(HookSupport.java:26)
	at net.jqwik.engine.hooks.statistics.StatisticsHook.aroundProperty(StatisticsHook.java:37)
	at net.jqwik.engine.execution.lifecycle.HookSupport.lambda$wrap$1(HookSupport.java:31)
	at net.jqwik.engine.execution.lifecycle.HookSupport.lambda$wrap$0(HookSupport.java:26)
	at net.jqwik.engine.hooks.lifecycle.AutoCloseableHook.aroundProperty(AutoCloseableHook.java:13)
	at net.jqwik.engine.execution.lifecycle.HookSupport.lambda$wrap$1(HookSupport.java:31)
	at net.jqwik.engine.execution.PropertyMethodExecutor.executePropertyMethod(PropertyMethodExecutor.java:113)
	at net.jqwik.engine.execution.PropertyMethodExecutor.execute(PropertyMethodExecutor.java:46)
	at net.jqwik.engine.execution.PropertyTaskCreator.executeTestMethod(PropertyTaskCreator.java:137)
	at net.jqwik.engine.execution.PropertyTaskCreator.lambda$createTask$1(PropertyTaskCreator.java:53)
	at net.jqwik.engine.execution.pipeline.ExecutionTask$1.lambda$execute$0(ExecutionTask.java:31)
	at net.jqwik.engine.execution.lifecycle.CurrentTestDescriptor.runWithDescriptor(CurrentTestDescriptor.java:17)
	at net.jqwik.engine.execution.pipeline.ExecutionTask$1.execute(ExecutionTask.java:31)
	at net.jqwik.engine.execution.pipeline.ExecutionPipeline.runToTermination(ExecutionPipeline.java:82)
	at net.jqwik.engine.execution.JqwikExecutor.execute(JqwikExecutor.java:46)
	at net.jqwik.engine.JqwikTestEngine.executeTests(JqwikTestEngine.java:70)
	at net.jqwik.engine.JqwikTestEngine.execute(JqwikTestEngine.java:53)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy5.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.lang.NullPointerException
	at java.base/java.util.Arrays.stream(Arrays.java:5614)
	at net.jqwik.engine.properties.shrinking.ErrorEquivalenceChecker.firstRelevantStackTraceElement(ErrorEquivalenceChecker.java:38)
	at net.jqwik.engine.properties.shrinking.ErrorEquivalenceChecker.areEquivalent(ErrorEquivalenceChecker.java:31)
	at net.jqwik.engine.properties.shrinking.PropertyShrinker.isFalsifiedButErrorIsNotEquivalent(PropertyShrinker.java:137)
	at net.jqwik.engine.properties.shrinking.PropertyShrinker.lambda$shrink$0(PropertyShrinker.java:55)
	at net.jqwik.engine.properties.shrinking.PropertyShrinker.lambda$shrinkAsLongAsSampleImproves$5(PropertyShrinker.java:116)
	at net.jqwik.engine.properties.shrinking.AbstractSampleShrinker.lambda$falsify$5(AbstractSampleShrinker.java:91)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1134)
	at net.jqwik.engine.properties.shrinking.AbstractSampleShrinker.falsify(AbstractSampleShrinker.java:91)
	at net.jqwik.engine.properties.shrinking.AbstractSampleShrinker.lambda$shrink$2(AbstractSampleShrinker.java:53)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.stream.ReferencePipeline$11$1.accept(ReferencePipeline.java:442)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(StreamSpliterators.java:302)
	at java.base/java.util.stream.Streams$ConcatSpliterator.tryAdvance(Streams.java:720)
	at java.base/java.util.stream.Streams$ConcatSpliterator.tryAdvance(Streams.java:723)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findAny(ReferencePipeline.java:548)
	at net.jqwik.engine.properties.shrinking.AbstractSampleShrinker.shrink(AbstractSampleShrinker.java:63)
	at net.jqwik.engine.properties.shrinking.OneAfterTheOtherParameterShrinker.shrinkSingleParameter(OneAfterTheOtherParameterShrinker.java:43)
	at net.jqwik.engine.properties.shrinking.OneAfterTheOtherParameterShrinker.shrink(OneAfterTheOtherParameterShrinker.java:25)
	at net.jqwik.engine.properties.shrinking.ShrinkingAlgorithm.shrinkOneParameterAfterTheOther(ShrinkingAlgorithm.java:52)
	at net.jqwik.engine.properties.shrinking.ShrinkingAlgorithm.shrink(ShrinkingAlgorithm.java:32)
	at net.jqwik.engine.properties.shrinking.PropertyShrinker.shrinkAsLongAsSampleImproves(PropertyShrinker.java:129)
	at net.jqwik.engine.properties.shrinking.PropertyShrinker.lambda$shrink$3(PropertyShrinker.java:86)
	at net.jqwik.engine.execution.lifecycle.CurrentTestDescriptor.runWithDescriptor(CurrentTestDescriptor.java:17)
	at net.jqwik.engine.properties.shrinking.PropertyShrinker.lambda$withTimeout$4(PropertyShrinker.java:98)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1692)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```
</details>

If you're curious for more context where this was blowing up in my code, you can see it here:
[scottg489/conjob:/src/test/unit/java/conjob/resource/admin/task/DockerVolumeRemoveTaskTest.java#L56](https://github.com/ScottG489/conjob/blob/master/src/test/unit/java/conjob/resource/admin/task/DockerVolumeRemoveTaskTest.java#L56)

Where that line was previously:
`doThrow(mockException).when(mockDockerAdapter).removeVolume(any());`
And since it was always throwing an error, given how the [prod code](https://github.com/ScottG489/conjob/blob/ae8cebb41136114e3eee1d566590829ed5004121/src/main/java/conjob/resource/admin/task/DockerVolumeRemoveTask.java#L23) worked it ended up bubbling up.

**Comment by jlink** [comment]
> I think I'm still siding towards the lib being able to handle everything gracefully. Maybe it's too hard of a line, but I don't think it should matter what the user is doing, the framework should not blow up.

I fully agree. The crucial thing is what _blow up_ means. I could argue that jqwik does not blow up, since it reports the one property as failing. Everything else continues to work.

As for printing out the mockException that produces the RTE, that's not possible without introducing code at the point of accessing the potential null attribute. And that would lead back to handling every potential null access explicitly. 

That being said, I now see a reason why in this case a special handling for null is warranted: `ErrorEquivalenceChecker` is a component that's crucial for the core functionality of shrinking and thus it should never fail in possible circumstances. And clearly it is possible to sneak in a throwable with null stacktrace.

So you win :-) I'll fix it.

**Comment by jlink** [comment]
Fixed and deployed to 1.6.3-SNAPSHOT (https://jqwik.net/docs/current/user-guide.html#snapshot-releases)
